### DISCUSSION TITLE
RandomVariables raise ValueError when size would broadcast parameter

### DISCUSSION THREAD

**Discussion by ricardoV94** [aesara-devs/aesara#1044] [General]
We have a common check in multivariate distributions (+ categorical), that raises when a non-size is not strictly equal to the batched dimensions of the parameters, saying objects cannot be broadcast to a single shape.

https://github.com/aesara-devs/aesara/blob/510a9618454293fc00b3cfa78b337c1f2d6d2af3/aesara/tensor/random/basic.py#L377-L383

I don't know why we are imposing this limitation. Even if there is a good reason, the message is wrong, because the batched shapes may very well be broadcastable to size:

```python
# Raises ValueError: shape mismatch: objects cannot be broadcast to a single shape
# However batch_shape=(1,) and size=(5,) are clearly broadcastable
at.random.dirichlet([[0.2, 0.3, 0.5]], size=5).eval()
```

The message for the CategoricalRV is more accurate:

https://github.com/aesara-devs/aesara/blob/510a9618454293fc00b3cfa78b337c1f2d6d2af3/aesara/tensor/random/basic.py#L651-L652

However I still don't get why we would not broadcast the parameter to match size, when this is would be valid. Otherwise, the following should also be invalid, no?

```python
# Just fine
at.random.normal([0], size=5).eval()
```

**Comment by brandonwillard** [comment]
You can check the `git blame` results at those locations and find the relevant commit(s), and&mdash;through that&mdash;the associated PR: https://github.com/aesara-devs/aesara/pull/446.

**Reply by ricardoV94** [reply]
I know that it came through that PR. It does not explain the interpretation with respect to broadcasting to size.

It explains why you cannot do `at.random.dirichlet(np.ones((2, 5)), size=5)`, the same way one cannot do `at.random.normal(np.ones((2,)), size=5)`

It does not explain why parameters can broadcast to size in univariate and not in multivariate distributions. Also the categorical distribution is a univariate one, and has the same limitation as other multivariate distributions.

**Reply by brandonwillard** [reply]
> It does not explain why parameters can broadcast to size in univariate and not in multivariate distributions.

For completeness, they _can_:
```python
import aesara.tensor as at

at.random.dirichlet([[0.2, 0.3, 0.5], [0.5, 0.3, 0.2]], size=(4, 2)).eval().shape
# (4, 2, 3)
```
Again, what you were observing was just a bug.

**Reply by brandonwillard** [reply]
> Also the categorical distribution is a univariate one, and has the same limitation as other multivariate distributions.

The categorical should work, too:
```python
at.random.categorical([[0.2, 0.3, 0.5], [0.5, 0.3, 0.2]], size=(4, 2,)).eval().shape
# (4, 2)
```

**Comment by ricardoV94** [comment]
I disagree with converting this to a discussion. The original example clearly shows a different interpretation/inconsistency between the meaning of size in univariate and multivariate distributions.

These checks were introduced by the PR that intended to harmonize the meaning across the two classes of distributions.

It also shows the distinct interpretation being used for the univariate categorical distribution that just so happens to require a vector parameter.

It looks like (the independent dimensions of parameters) should be allowed to broadcast to size. If not, the last example with the normal should also fail (and similar checks should be introduced to all existing univariate distributions).

**Reply by brandonwillard** [reply]
Do not convert this into another issue until some sort of resolution is arrived upon via this discussion.

~If you read the original PR (i.e. https://github.com/aesara-devs/aesara/pull/446) and the associated discussion in https://github.com/aesara-devs/aesara/discussions/412, you'll see that the current approach is well justified.  If you believe that there are benefits to a different approach, or errors/issues with the current, then you need to make the relevant comparisons and justifications clear in this discussion.~

~As of now, you haven't addressed the consistency issue that was explicitly addressed by the current approach, especially since what you're proposing is only a mild convenience in some cases and introduces an important **inconsistency**: it implements broadcasting **in the opposite direction** as the standard NumPy broadcasting rules.~

~More specifically, your example in #1056, i.e.~
```python
import numpy as np
import aesara.tensor as at

# Already allowed before
at.random.normal(np.ones((3, 1, 3)), size=(3, 10, 3)).eval().shape  # (3, 10, 3)

# Allowed now
at.random.dirichlet(np.ones((3, 1, 3)), size=(3, 10)).eval().shape  # (3, 10, 3)
```
~it proposes that an array of shape `(3, 1, 3)` (i.e. matching the independent parameters given by `np.ones((3, 1, 3))`) should "broadcast" against the "size" `(3, 10)`, but that's not how broadcasting works.  Broadcasting goes from the rightmost dimensions&mdash;of both shapes&mdash;to the left, adding broadcast dimensions wherever one array is missing a dimension that the other has.~  

~Your proposed change implements broadcasting in the opposite direction, and this is an inconsistency we don't need.  The only way to reconcile this inconsistency is by saying that `size` works differently for multivariate distributions, but this is the kind of "special-casing" and unnecessary complexity that https://github.com/aesara-devs/aesara/discussions/412 was addressing.~

I thought your "Already allowed before" example was somehow implying that the `at.random.normal` case should work with `size=(3, 10)` as well, like the `dirichlet` example that follows.

> The original example clearly shows a different interpretation/inconsistency between the meaning of size in univariate and multivariate distributions.

The original example looks like it was just a _bug_, and shouldn't have anything to do with the interpretation/meaning of `size` between univariate and multivariate distributions.

**Reply by brandonwillard** [reply]
Wait, if you're only talking about broadcasting the dimensions _other than the support dimension_, then this might not be an issue, because it would be in-line with https://github.com/aesara-devs/aesara/pull/446 and https://github.com/aesara-devs/aesara/discussions/412.

**Comment by brandonwillard** [comment]
> Even if there is a good reason, the message is wrong, because the batched shapes may very well be broadcastable to size:
> 
> ```python
> # Raises ValueError: shape mismatch: objects cannot be broadcast to a single shape
> # However batch_shape=(1,) and size=(5,) are clearly broadcastable
> at.random.dirichlet([[0.2, 0.3, 0.5]], size=5).eval()
> ```

This could simply be a bug in the current implementation of `RandomVariable`, and not a general issue with the way `size` is interpreted, because&mdash;for instance&mdash;`at.random.dirichlet([[0.2, 0.3, 0.5]], size=(5,)).shape.eval()` yields `[5, 3]`.  Also, `at.random.dirichlet([[0.2, 0.3, 0.5], [0.2, 0.3, 0.5]], size=(2,)).eval()` works, so it seems to be a problem specific to _broadcastable independent dimensions_.

**Reply by brandonwillard** [reply]
Now, that the original example issue is resolved by https://github.com/aesara-devs/aesara/pull/1056, it's clear that the problem was specific to the same copy-pasted code used by `MvNormalRv.rng_fn`, `MultinomialRV.rng_fn`, and `DirichletRV.rng_fn`, and not our handling of multivariate distribution parameters in general (e.g. in the `RandomVariable` methods or anywhere else).
